### PR TITLE
Migrating precommit hooks from LB

### PR DESCRIPTION
**Pull request by jaredponn**
Builds off of: https://github.com/mlabs-haskell/flake-lang.nix/pull/18

- Removed `./examples/pre-commit.nix` in favour of augmenting
  `./pre-commit.nix`
   - Justification: I'd prefer to keep all the `pre-commit` things in a
     single file.

- Added a `flakeModule.rustMonorepoPreCommit` attribute which contains a
  module for formatting rust code
    - Justification: This copies the API given in
      https://github.com/cachix/pre-commit-hooks.nix/blob/7c54e08a689b53c8a1e5d70169f2ec9e2a68ffaf/flake.nix#L27.
      My reading of this is that we want to expose a flake module which
      enables the rust formatter. @szg251 if we're happy with this, I
      can try to get this patched in `pre-commit-hooks.nix` as a more
      permanent solution.

- Removed deno precommit hooks
    - Justification: originally we rolled our own `deno` precommit
      things because the default was slightly broken. But it has now
      been patched -- see https://github.com/cachix/pre-commit-hooks.nix/pull/377

- Ran precommit hooks for TS

### CODE REVIEW

File: examples/pre-commit.nix [hunk 1]
```diff
@@ -1,15 +0,0 @@
-{ inputs, config, ... }: {
-  imports = [
-    inputs.pre-commit-hooks.flakeModule
-    config.flake-lang.rustMonorepoPreCommit
-    config.flake-lang.denoPreCommit
```
Comment (jaredponn) [inline examples/pre-commit.nix:5]: Just a remark -- I don't think we can do this i.e., import a value set in the `config`. In Slack, I've cooked up a minimal example to demonstrate.

File: examples/pre-commit.nix [hunk 2]
```diff
@@ -1,15 +0,0 @@
-{ inputs, config, ... }: {
```
Comment (jaredponn) [inline examples/pre-commit.nix:1]: I moved all of this to `./pre-commit.nix` to keep all code quality things in one place

File: flake-lang/build.nix [hunk 3]
```diff
@@ -81,22 +104,6 @@
...
-          };
-
-          denoPreCommit = lib.mkOption {
-            type = lib.types.flakeModule;
-            default = ./pre-commit-hooks/deno.nix;
-            readOnly = true;
-            description = lib.mdDoc ''pre-commit-hooks.nix hook for Deno in a monorepo setting'';
-            example = lib.mdDoc ''TODO(bladyjoker)'';
-          };
-
```
Comment (jaredponn) [inline flake-lang/build.nix:99]: We don't need to roll our own deno anymore. They fixed this in the main repo

File: examples/typescript-flake-project-with-extra-dependency/package.json [hunk 4]
```diff
@@ -18,10 +18,7 @@
   },
   "author": "Jared Pon",
```
Comment (jaredponn) [inline examples/typescript-flake-project-with-extra-dependency/package.json:19]: Ignore changes in this file: this is just precommit hooks

File: examples/typescript-flake-project-with-extra-dependency/src/MyLib/index.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
-import * as TsFlakeProject from 'typescript-flake-project'
+import * as TsFlakeProject from "typescript-flake-project";
```
Comment (jaredponn) [inline examples/typescript-flake-project-with-extra-dependency/src/MyLib/index.ts:1]: Ignore changes in this file: this is just precommit hooks

File: examples/typescript-flake-project-with-transitive-extra-dependency/src/MyLib/index.ts [hunk 6]
```diff
@@ -1,9 +1,7 @@
-import * as TsFlakeProject from 'typescript-flake-project'
```
Comment (jaredponn) [inline examples/typescript-flake-project-with-transitive-extra-dependency/src/MyLib/index.ts:1]: Ignore changes in this file: this is just precommit hooks

File: examples/typescript-flake-project/package.json [hunk 7]
```diff
@@ -18,10 +18,7 @@
   },
```
Comment (jaredponn) [inline examples/typescript-flake-project/package.json:18]: Ignore changes in this file: this is just precommit hooks

File: examples/typescript-flake-project/src/MyLib/index.ts [hunk 8]
```diff
@@ -1,3 +1,3 @@
 export function helloWorld() {
-    console.log("Hello world!");
```
Comment (jaredponn) [inline examples/typescript-flake-project/src/MyLib/index.ts:2]: Ignore changes in this file: this is just precommit hooks

File: flake-lang/build.nix (hunk 3 above)
Comment (jaredponn) [reply flake-lang/build.nix:99]: Although this prompts some discussion. In Rust, it seems we want to export a module which does all the formatting -- do we want this for all languages (or none)?

File: flake-lang/build.nix [hunk 9]
```diff
@@ -16,11 +19,31 @@
...
         visible = false;
       };
+
+      # Expose attributes like
+      # ```
+      # flakeModule.rustMonorepoPreCommit = ..
+      # ```
+      flakeModule = {
+        rustMonorepoPreCommit = lib.mkOption {
+          type = lib.types.lazyAttrsOf lib.types.unspecified;
```
Comment (bladyjoker) [inline flake-lang/build.nix:39]: Should this be a `flakeModule` type if there is such a thing?

File: flake-lang/build.nix (hunk 3 above)
Comment (bladyjoker) [reply flake-lang/build.nix:99]: Not sure what you mean...

AFAIK this is about providing Rust formatting with pre-commit-checks.nix, there were reasons why we had to roll our own. @szg251 any comments?

File: flake-lang/pre-commit-hooks/tools.nix
Comment (bladyjoker) [inline flake-lang/pre-commit-hooks/tools.nix:1]: This is very lovely, thanks!

**Review by bladyjoker — APPROVED** [review]
Awesome thanks

File: examples/pre-commit.nix (hunk 1 above)
Comment (bladyjoker) [reply examples/pre-commit.nix:5]: That makes a lot of sense!

**Pull request merged**